drivers/sensor/icm42688/icm42688_decoder.c

```diff
@@ -0,0 +1,251 @@
+		 needle == SENSOR_CHAN_ACCEL_Z)) ||
+	       (haystack == SENSOR_CHAN_GYRO_XYZ &&
+		(needle == SENSOR_CHAN_GYRO_X || needle == SENSOR_CHAN_GYRO_Y ||
+		 needle == SENSOR_CHAN_GYRO_Z)) ||
+	       (haystack == SENSOR_CHAN_MAGN_XYZ &&
+		(needle == SENSOR_CHAN_MAGN_X || needle == SENSOR_CHAN_MAGN_Y ||
+		 needle == SENSOR_CHAN_MAGN_Z));
+}
+
+static int get_shift(const uint8_t *buffer, enum sensor_channel channel_type, int8_t *shift)
+{
+	struct sensor_data_generic_header *header = (struct sensor_data_generic_header *)buffer;
+
+	for (uint8_t i = 0; i < header->num_channels; ++i) {
+		if (is_subchannel(channel_type, header->channel_info[i].channel)) {
+			*shift = header->channel_info[i].shift;
+			return 0;
+		}
+	}
+	return -EINVAL;
+}
+
+static int compute_channel_info_index(const struct sensor_data_generic_header *header,
+				      sensor_channel_iterator_t cit, int *index, int *offset)
+{
+	sensor_channel_iterator_t i = {0};
+
+	*index = 0;
+	do {
+		sensor_channel_iterator_t last_i = i;
+
+		i += SENSOR_CHANNEL_3_AXIS(header->channel_info[*index].channel) ? 3 : 1;
+
+		if (i > cit) {
+			*offset = cit - last_i;
+			return 0;
+		}
+		*index += 1;
+	} while (*index < header->num_channels);
+
+	return -EINVAL;
+}
+
+static int decode(const uint8_t *buffer, sensor_frame_iterator_t *fit,
+		  sensor_channel_iterator_t *cit, enum sensor_channel *channels, q31_t *values,
+		  uint8_t max_count)
+{
+	const struct sensor_data_generic_header *header =
+		(const struct sensor_data_generic_header *)buffer;
+	int num_samples = 0;
+	const icm42688_sample_t *q =
+		(const icm42688_sample_t *)(buffer + sizeof(struct sensor_data_generic_header) +
+					    header->num_channels *
+						    sizeof(struct sensor_data_generic_channel));
+
+	if (*fit != 0) {
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < header->num_channels; ++i) {
+		num_samples += SENSOR_CHANNEL_3_AXIS(header->channel_info[i].channel) ? 3 : 1;
+	}
+
+	int count = 0;
+	int channel_info_index;
+	int offset;
+
+	if (compute_channel_info_index(header, *cit, &channel_info_index, &offset) != 0) {
+		LOG_ERR("Could not compute the channel index");
+		return -EINVAL;
+	}
+
+	for (; *cit < num_samples && count < max_count; ++count) {
+		enum sensor_channel channel = header->channel_info[channel_info_index].channel;
+
+		channels[count] = channel;
+		if (SENSOR_CHANNEL_3_AXIS(channel)) {
+			channels[count] = channel - 3 + offset;
+			offset++;
+			if (offset == 3) {
+				channel_info_index++;
+				offset = 0;
+			}
+		} else {
+			channel_info_index++;
+		}
+		values[count] = q[*cit];
+		*cit += 1;
+	}
+
+	if (*cit >= num_samples) {
+		*fit += 1;
+		*cit = 0;
+	}
+	return 0;
+}
+
+struct sensor_decoder_api icm42688_decoder = {
+	.get_frame_count = get_frame_count,
+	.get_timestamp = get_timestamp,
+	.get_shift = get_shift,
+	.decode = decode,
+};
+
+int icm42688_get_decoder(const struct device *dev, struct sensor_decoder_api *decoder)
+{
+	*decoder = icm42688_decoder;
```

Comment:
Is this supposed to copy the decoder struct or did you mean for `decoder` to be type `struct sensor_decoder_api **` and do `*decoder = &icm42688_decoder;`?

Reply:
This matches the API Yuval provided in `sensor.h`
```
typedef int (*sensor_get_decoder_t)(const struct device *dev,
				    struct sensor_decoder_api *api);
```